### PR TITLE
Insert annotations at the correct position for arrays

### PR DESCRIPTION
**Pull request by avenger2597**
This PR helps insert annotations at the correct position for array declarations.

For array declarations we want the annotations to be inserted just before the array type brackets. For example : 
String `@Nullable [] f2;`

Currently, the behaviour is that the annotations are inserted before the simpleName type. Like this :
`@Nullable String[] f2;`

### CODE REVIEW

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnLocalVariableInjectionTest.java [hunk 1]
```diff
@@ -244,9 +244,9 @@ public void onArrayType() {
             "import edu.ucr.UnTainted;",
             "public class Foo<T> {",
             "   public void foo() {",
-            "      @UnTainted Map<String, T[]>[] f0;",
```
Comment (nimakarimipour) [inline injector/src/test/java/edu/ucr/cs/riple/injector/OnLocalVariableInjectionTest.java:247]: This failure is a bug in your changes not that the expected output is wrong.

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java [hunk 2]
```diff
@@ -62,7 +62,17 @@ Modification computeTextModificationOn(T node) {
     if (annotAlreadyExists) {
       return null;
     }
-    return new Insertion(annotationExpr.toString(), range.begin);
+
+    String annotation = annotationExpr.toString();
```
Comment (nimakarimipour) [inline injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java:66]: I don't think this is a right place to add your changes. Your changes should alter [AddTypeUseMarkerAnnotation](https://github.com/ucr-riple/NullAwayAnnotator/blob/master/injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java)  behavior. 

**Hint**: all you changes should be in [here](https://github.com/ucr-riple/NullAwayAnnotator/blob/c965c90891c181c29a21c4bf8731946422be4f73/injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java#L611) where we are returning array component range.

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java [hunk 3]
```diff
@@ -62,6 +62,35 @@ public void fieldNullableSimple() {
         .start();
   }
 
+  @Test
+  public void fieldNullableArray() {
```
Comment (nimakarimipour) [inline injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java:66]: This is not a good name for this test. Unit test names should demonstrate what feature we are testing exactly.

**Comment by nimakarimipour** [comment]
@avenger2597 Make sure you merge the latest changes on master branch before resolving comments.

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java [hunk 4]
```diff
@@ -62,7 +62,17 @@ Modification computeTextModificationOn(T node) {
...
+    // Check node for array
+    int index = node.toString().indexOf('[');
+    // Node has array
+    if (index != -1) {
+      int begin = range.begin.column + index;
```
[the remark this answers is not in the document]
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java]: Updated this

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java [hunk 5]
```diff
@@ -62,6 +62,35 @@ public void fieldNullableSimple() {
...
+    injectorTestHelper
+        .addInput(
+            "Foo.java",
+            "package test;",
+            "import javax.annotation.Nullable;",
+            "public class Foo {",
+            "   Object[] h = new Object[4];",
+            "   public void test(@Nullable Object[] f) {",
```
[the remark this answers is not in the document]
Comment (avenger2597) [reply injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java]: Removed this test case entirely.  Currently using TypeUseAnnotationTest.additionOnArrayTest() to test my chages. Can add more if needed later.

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java [hunk 6]
```diff
@@ -62,7 +62,17 @@ Modification computeTextModificationOn(T node) {
     if (annotAlreadyExists) {
       return null;
     }
-    return new Insertion(annotationExpr.toString(), range.begin);
+
+    String annotation = annotationExpr.toString();
+    // Check node for array
+    int index = node.toString().indexOf('[');
```
[the remark this answers is not in the document]
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java]: Updated

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnLocalVariableInjectionTest.java (hunk 1 above)
Comment (avenger2597) [reply injector/src/test/java/edu/ucr/cs/riple/injector/OnLocalVariableInjectionTest.java]: Reverted to original

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java [hunk 7]
```diff
@@ -62,6 +62,35 @@ public void fieldNullableSimple() {
...
+            "package test;",
+            "import javax.annotation.Nullable;",
+            "public class Foo {",
+            "   Object @Nullable [] h = new Object[4];",
+            "   public void test(@Nullable Object[] f) {",
+            "      h = f;",
+            "   }",
+            "}")
+        .addChanges(
+            new AddMarkerAnnotation(
```
[the remark this answers is not in the document]
Comment (avenger2597) [reply injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java]: Removed this test case.

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java (hunk 2 above)
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddMarkerAnnotation.java]: Done

File: injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java (hunk 3 above)
Comment (avenger2597) [reply injector/src/test/java/edu/ucr/cs/riple/injector/OnFieldInjectionTest.java]: Removed this test case

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java [hunk 8]
```diff
@@ -68,6 +69,13 @@ public Modification computeTextModificationOnType(Type type, AnnotationExpr anno
     if (range == null) {
       return null;
     }
+
+    if (type instanceof ArrayType) {
```
Comment (nimakarimipour) [inline injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java:73]: @avenger2597 As mentioned previously in this [comment](https://github.com/ucr-riple/NullAwayAnnotator/pull/251#discussion_r1795932716) your changes should be in `findSimpleNameRangeInTypeName` method where we are returning the array's component type as the place to insert / remove annotation for arrays. You should just change the existing case for array and use `type.getTokenRange()` to compute the range right before the first `[`.

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java [hunk 9]
```diff
@@ -68,6 +69,13 @@ public Modification computeTextModificationOnType(Type type, AnnotationExpr anno
...
+      ArrayType.Origin origin = ((ArrayType) type).getOrigin();
+      if (origin == ArrayType.Origin.TYPE) {
```
Comment (nimakarimipour) [inline injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java:75]: Let's not worry about this yet.

**Review by nimakarimipour — CHANGES_REQUESTED** [review]

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java (hunk 9 above)
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java]: Removed

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java (hunk 8 above)
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java:74]: Done.

File: injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java [hunk 10]
```diff
@@ -608,7 +609,14 @@ public static Range findSimpleNameRangeInTypeName(Type type) {
       return ((ClassOrInterfaceType) type).getName().getRange().get();
     }
     if (type instanceof ArrayType) {
-      return findSimpleNameRangeInTypeName(((ArrayType) type).getComponentType());
+      Optional<TokenRange> tokenRange = type.getTokenRange();
+      int index = type.toString().indexOf('[');
```
Comment (nimakarimipour) [inline injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java:613]: This is not the right way to compute the index. The `type.toString()` will return string representation of the type, not how it is written in the source code. You should find the index via:
```
for (JavaToken javaToken : range) {
   // Token T = javaToken.asString();
   // Your logic
}
```

**Review by nimakarimipour — CHANGES_REQUESTED** [review]
One comment.

Also please write a single and simple unit test for this.

File: injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java (hunk 10 above)
Comment (avenger2597) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java]: Updated to this logic

File: injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java (hunk 8 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/changes/AddTypeUseMarkerAnnotation.java:74]: @avenger2597 Nice, this is much cleaner now.

File: injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java (hunk 10 above)
Comment (nimakarimipour) [reply injector/src/main/java/edu/ucr/cs/riple/injector/Helper.java]: @avenger2597 Nice, this is much cleaner now.

**Review by nimakarimipour — APPROVED** [review]

**Pull request merged**